practical/ProcessMining/group2/inductive_miner/src/algo.py

```diff
@@ -0,0 +1,527 @@
+
+        :param event_log: EventLog object.
+        """
+        self.event_log = event_log
+        self.graph = defaultdict(list)  # Adjacency list, format: "node_id" : [children]
+        self.start_nodes = set()
+        self.end_nodes = set()
+
+    def construct_dfg(self) -> None:
+        """Construct Directly Follows Graph."""
+        for trace in self.event_log.traces.keys():
+            if trace:  # Check if trace is not empty
+                self.start_nodes.add(trace[0])  # First activity in trace is a start
+                self.end_nodes.add(trace[-1])  # Last activity in trace is an end
+
+                for i in range(len(trace) - 1):
+                    current_activity = trace[i]
+                    next_activity = trace[i + 1]
+
+                    if next_activity not in self.graph[current_activity]:
+                        self.graph[current_activity].append(next_activity)
+                    if (next_activity not in self.graph):  # Add next activity to graph if not already present, i.e. if it is an end node
+                        self.graph[next_activity] = []
+                
+                for node in self.start_nodes:
+                    if node not in self.graph: # Isolated nodes
+                        self.graph[node] = []
+            else:
+                self.graph[''] = [] # Add tau node for empty traces
+
+    def __str__(self) -> str:
+        """Represent Directly Follows Graph as string."""
+        return f"Directly Follows Graph: (\n\tGraph: {dict(self.graph)}\n\tStart nodes: {self.start_nodes}\n\tEnd nodes: {self.end_nodes}\n)"
+
+class ProcessTree:
+    def __init__(self, event_log: EventLog) -> None:
+        """
+        Initialize ProcessTree object.
+
+        :param event_log: EventLog object.
+        """
+        self.event_log = event_log
+        self.root = None
+        self.children = []
+
+    def find_base_case(self) -> str:
+        """
+        Find base case for the Process Tree.
+
+        :return: Base case activity or None.
+        """
+        if len(self.event_log.traces) == 0:
+            return 'tau'
+        elif len(self.event_log.traces) == 1:
+            only_trace = next(iter(self.event_log.traces))
+            if only_trace == "":
+                return 'tau'
+            elif len(only_trace) == 1:
+                return only_trace
+        return None
+
+    def find_exclusive_choice_cut(self, dfg: DirectlyFollowsGraph) -> List[List[str]]:
+        """
+        Find exclusive choice cut.
+
+        :param dfg: DirectlyFollowsGraph object.
+        :return: List of lists representing cuts.
+        """
+        # Convert the graph to undirected
+        undirected = dfg.convert_to_undirected()
+        # Find connected components
+        cuts = undirected.find_components()
+
+        return None if len(cuts) == 1 else cuts
+    
+    def find_sequence_cut(self, dfg: DirectlyFollowsGraph) -> List[List[str]]:
+        """
```

Comment:
Sequence cuts logic is nasty, but refactoring it into seperate methods for each step could improve readability

Reply:
Agree, but abandoned due to time and sanity reasons